src/sentry/api/endpoints/debug_files.py

```diff
@@ -63,6 +63,12 @@ def has_download_permission(request, project):
     organization = project.organization
     required_role = organization.get_option("sentry:debug_files_role") or DEBUG_FILES_ROLE_DEFAULT
 
+    if request.user.is_sentry_app:
+        if roles.get(required_role).priority > roles.get("member").priority:
+            return request.access.has_scope("project:write")
```

Comment:
Note that this is a hack. I would love @getsentry/ecosystem to take a look at this and other places in a follow-up to determine what a proper authentication model for integrations would look like in such a case.

Reply:
@jan-auer There are a few places where we bypass a check on the existence or role of an `OrganizationMember` for the integration platform. Most of them look pretty similar to this. I think the best we could do is create some functions like `user_has_scope_for_org` that would abstract the logic checking whether it's a sentry app, and if not, query the `OrganizationMember` table.

One thing I do want to point out is that the check that was added is not actually checking that the integration is actually installed on the org that you're checking. It might be OK in that we check this upstream in the `ProjectPermission` class where this is used but it does mean if `has_download_permission` is used somewhere else we could have a vulnerability.

Reply:
> It might be OK in that we check this upstream in the ProjectPermission class where this is used

This is what we were counting on in this function, as we didn't want to run every check twice. This is a very valid concern, though, and now I think this deserves at least a code comment explaining why we don't check for installation.

Ultimately, our challenge centers around how roles and scopes are exposed in the product:
- Roles are given to users in organizations and define scopes they come with. However, which role owns which scope is not transparent to the user, and it is mostly irrelevant. An organization owner adding people and configuring access rights likely has a concept of roles ("I want only Admins to access that"), but not about scopes.
- Scopes are the primary means of access control. In addition to that, API tokens and integration tokens specify which scopes they receive. Our product documentation often declares which scopes are needed for certain operations, but this is not consistent. The API documentation consistently lists this.

Our objective is to fine-tune access control for certain sensitive entities. At the moment, these are Attachments and Debug Files. Most organizations will be fine if everyone can access them, but some organizations may need to further restrict access.

Let's focus on `project:write` vs `project:read` for a moment. Every role higher than Member automatically gets `project:write` for every project in the organization. We have the following options:
1. Configure the required scope. This makes it directly compatible with all tokens, but effectively means you can no longer differentiate between Admins, Managers, and Owners, since they all have this scope.
2. Configure the role. This is no longer with the integration platform and API tokens. The current hack is that we map every role other than Member to `project:write`, which is slightly more restrictive, but effectively one can circumvent e.g. an "Owner" restriction with this. 
3. Build explicit access control, where users can specify either other users or teams. Integrations could be added to teams, in which case they receive access to the team's resources. The default could be all teams. This would probably be the most natural way of doing this, similar to how GitHub works. It is quite a big refactor, however.

Reply:
@jan-auer That's an excellent writeup, thanks for that.
> This is what we were counting on in this function, as we didn't want to run every check twice. This is a very valid concern, though, and now I think this deserves at least a code comment explaining why we don't check for installation.

I'm somewhat inclined to say that I rather have a function that runs inexpensive checks multiple times if it means the encapsulated abstraction doesn't have logical holes that could lead us open to a vulnerability with a future code change.

> 2. Configure the role. This is no longer with the integration platform and API tokens. The current hack is that we map every role other than Member to project:write, which is slightly more restrictive, but effectively one can circumvent e.g. an "Owner" restriction with this.

I think this is probably the best solution. We can have something that maps the scopes of integration to the equivalent user role:
1. `project:write` -> Manager (I would argue that something that requires an Owner role should _not_ be automated but maybe that's an unnecessary protection)
2.  `project:read` -> Member
3. No project scope -> `None` (not a member)

We can have a function on the user class `get_role_for_organization` to do this and it can abstract away all the differences between an integraion proxy user and a normal user.

Reply:
> I'm somewhat inclined to say that I rather have a function that runs inexpensive checks multiple times if it means the encapsulated abstraction doesn't have logical holes that could lead us open to a vulnerability with a future code change.

Agreed, the cost of a second database query (is it?) is worth the added security in the future.

> I would argue that something that requires an Owner role should not be automated but maybe that's an unnecessary protection

Excellent point. Would that mean that we reject all access from automation tokens when "Owner" is chosen? This means the mapping from the setting would be:

- _Member_ -> require `project:read`
- _Admin_ -> require `project:write`
- _Manager_ -> require `project:write` (or reject?)
- _Owner_ -> reject

> We can have a function on the user class get_role_for_organization to do this and it can abstract away all the differences between an integration proxy user and a normal user.

👍

Reply:
Driving by since we might do something similar for alerts soon: since it's not clear what the role selection means for auth tokens I would call it out in the subtext of the setting